content/blog/typestate-pattern/index.md

```diff
@@ -0,0 +1,303 @@
++++
+title = "The Typestate Pattern in Rust"
+date = 2024-05-16
+template = "article.html"
+[extra]
+series = "Idiomatic Rust"
+reviews = []
++++
+
+Modern cars are essentially computers on wheels, containing a staggering amount of software. Did you know that a modern car has over [100 million lines of code](https://medium.com/next-level-german-engineering/porsche-future-of-code-526eb3de3bbe)? That's an order of magnitude more than the 14 million lines of code found in a Boeing 787 Dreamliner.
+
+With so much software, it's no surprise that cars have bugs. In fact, there are [a lot of bugs](https://www.wired.com/story/uber-self-driving-crash-arizona-ntsb-report/), which is very reassuring given that [microprocessors are responsible for critical functions like braking](https://blog.passwork.pro/car-hacking/).
+
+Now, I know nothing about car software, but at the end of the day, it's just software. And there are patterns and techniques that can be applied to make software more reliable and secure. One such pattern is the typestate pattern, which I'll be exploring in this article.
+
+## The State of a Car
+
+One of the most fundamental security features of a car is the door lock. A car door can be in one of two states:
+
+- **Locked**: The door is locked and cannot be opened.
+- **Unlocked**: The door is unlocked and can be opened from the inside or outside.
+
+## Representing States in Rust
+
+Consider the following scenario: if you pull the handle and the door is locked, it should not open. This logic, though simple, must be handled correctly in software, at least since all cars come with keyless entry systems by now. Suppose a car manufacturer decided to implement this logic in Rust (an in fact, a few might). How could they use Rust's type system to the fullest to prevent illegal states?
```

Comment:
```suggestion
Consider the following scenario: if you pull the handle and the door is locked, it should not open. This logic, though simple, must be handled correctly in software, at least since all cars come with keyless entry systems by now. Suppose a car manufacturer decided to implement this logic in Rust (and in fact, a few might). How could they use Rust's type system to the fullest to prevent illegal states?
```

Reply:
Thanks. Rewrote that part.